emscripten_strict

juj opened the pull request:
This adds a new `EMSCRIPTEN_STRICT` build mode, enabled either with `-s EMSCRIPTEN_STRICT=1` or with env. var `EMSCRIPTEN_STRICT=1`. That does the following things:
- The system-provided `#define EMSCRIPTEN` preprocessor value disappears, but code should use `#ifdef __EMSCRIPTEN__`.
- The linker flag `-s ERROR_ON_UNDEFINED_SYMBOLS` defaults to `1` instead of the previous default `0`. (#2714)
- Adds a new linker flag `-s ERROR_ON_MISSING_LIBRARIES=0/1`, which defaults to 0, but in `EMSCRIPTEN_STRICT` mode defaults to 1. (#4213)
- Removes `system/include/emscripten` from the set of system include directories in `EMSCRIPTEN_STRICT` mode, with the intent that users should do `#include <emscripten/foo.h>` to include Emscripten-related system headers (see also #4664)
- Stops autolinking to most `src/library_xx.js` files, but instead code should use one of the predefined library aliases to link to system libraries, e.g. `-lGL` for OpenGL or `-lSDL` for SDL 1.x etc., or by using `-lfoo.js` to link to them. (#2730)

The intent is that these all would become deprecated options, and people would be able to use `EMSCRIPTEN_STRICT` mode to try out ahead of time that their projects build without relying on deprecated features.

In this PR, building without `EMSCRIPTEN_STRICT` mode doesn't yet change any existing behavior, but all changes are opt in at this point (although with the wording being about the intent to deprecate)

Tried to get as much of the breaking changes in one go to highlight/group the different things that I'd like to change. Note that these are all more or less validation related, where it would be nicest to be as strict as possible by default, like native compilers are.

juj commented:
Need to run full suite before merging, but wanted to post this already for initial comments. Also needs a good heads up conversation on the mailing list, but let's first discuss here that the different deprecations are something that we want to agree on, and what would be the best way to bring those in. (In one go like this, or separately?)

kripken commented:
Overall this sounds good. Main concerns are, as you said, to not surprise people and discuss everything well in advance, and yeah, a lailing list post to point to here makes sense.

> This adds a new EMSCRIPTEN_STRICT build mode, enabled either with -s EMSCRIPTEN_STRICT=1 or with env. var EMSCRIPTEN_STRICT=1. 

How about `-s STRICT=1` for the option (seems odd to have `EMSCRIPTEN_` in an option), and `EMCC_STRICT=1` for the env var (others start with `EMCC`)?

juj commented:
Alright. In the first version/sdk release, let's land this in a manner that does not remove any features, but just adds the strict mode. And in a later release, do the actual removal. That would give the kind of "two stage deprecate->remove" process to ensure not too abrupt changes. I'll post to the mailing list about this one to raise awareness.

kripken commented on emcc.py:
```diff
@@ -250,7 +250,10 @@ def filter_emscripten_options(argv):
+    if not use_js:
+      cmd += shared.EMSDK_OPTS + ['-D__EMSCRIPTEN__']
+      # The preprocessor define EMSCRIPTEN is deprecated. Don't pass it to code in strict mode. Code should use the define __EMSCRIPTEN__ instead.
+      if not os.environ.get('EMSCRIPTEN_STRICT') or int(os.environ.get('EMSCRIPTEN_STRICT')) == 0: cmd += ['-DEMSCRIPTEN']
```
would be nicer to have a variable `strict_mode` or such, that does the environment check once at the top of the file.

in fact maybe it could be in `tools/shared.py`, as others might need it?

juj replied:
Sounds good.

kripken commented on tools/shared.py:
```diff
@@ -931,8 +931,10 @@ def get_llvm_target():
+  COMPILER_OPTS = COMPILER_OPTS + ['-D__EMSCRIPTEN__']
+
+  # The preprocessor define EMSCRIPTEN is deprecated. Don't pass it to code in strict mode. Code should use the define __EMSCRIPTEN__ instead.
+  if not os.environ.get('EMSCRIPTEN_STRICT') or int(os.environ.get('EMSCRIPTEN_STRICT')) == 0: COMPILER_OPTS += ['-DEMSCRIPTEN']
```
ah yes, it is needed here. so maybe have `Building.is_strict()`, in this file, and we call it from all the places that need to ask if we are in strict mode?

juj replied:
Added `shared.is_emscripten_strict()`. Could not put it in `Building`, because that is too late in the file for early bits of the Settings-related code to refer to it (and reordering would create other kind of dependency issue, with Building referring to Settings too early), so made it a self-standing function at the top of the file.

kripken commented:
Meanwhile reading this a second time, I do see the logic in calling the option `EMSCRIPTEN_STRICT` - it's something about the behavior of emscripten itself, specifically. So I could go either way. Let's see what more people think.

dschuff commented:
+1, this is great; it should result in more consistent (and therefore hopefully less surprising) behavior, and allow us more implementation flexibility.
I think just `STRICT` is fine for the `-s` flag, and whatever is consistent with other env vars for that.

juj commented:
I think it helps to have the linker flag `-s EMSCRIPTEN_STRICT=1` coincide with the env. var name `export EMSCRIPTEN_STRICT=1` so that one wouldn't need to rememeber two separate names for the same feature?

kripken commented:
Well, this would be the first env var that has the exact same name as a setting - which opens up opportunities for confusion. Actually, though, why do we need both an env var *and* a setting here?

juj commented:
> Well, this would be the first env var that has the exact same name as a setting - which opens up opportunities for confusion. Actually, though, why do we need both an env var and a setting here?

I'd like to give as much flexibility as possible to be able to easily configure the feature to builds. Env. vars are great for locally/temporarily setting a value to try out once, but linker flags are better for build systems.

kripken commented:
Are you suggesting we temporarily support the two ways, or forever?

juj commented:
Ok, renamed to env. var `EMCC_STRICT=1` and `-s STRICT=1`. I'm suggesting that we do support this env. var and linker flag permanently, but that the features that they add are not permanent. That is, the changes that strict mode adds now, e.g. no `#define EMSCRIPTEN` etc. would not be fixed to be things only in strict mode, but in the future what is now strict mode would be the default mode, so in a sense use strict mode as a "landing pad" to introduce deprecations (when there is no other flag to do so).

kripken commented on emcc.py:
```diff
@@ -935,8 +940,39 @@ def detect_fixed_language_mode(args):
-          logging.warning('emcc: cannot find library "%s"', lib)
+        if not found:
+          # Some native libraries are implemented in Emscripten as system side JS libraries
+          js_system_libraries = {
```
how about refactoring this js system lib logic into `Building` in `tools/shared.py`?

kripken replied:
I agree it's not code likely to be used elsewhere. But just imagine adding a js library in the future, it feels nicer to edit a small function in `shared.py` rather than some code in the middle of `emcc.py`?

juj replied:
Ok, moved the code to `Building`.

kripken commented on tools/shared.py:
```diff
@@ -931,8 +935,10 @@ def get_llvm_target():
   # wasm target does not automatically define emscripten stuff, so do it here.
-  COMPILER_OPTS = COMPILER_OPTS + ['-DEMSCRIPTEN',
-                                   '-D__EMSCRIPTEN__']
+  COMPILER_OPTS = COMPILER_OPTS + ['-D__EMSCRIPTEN__']
```
looks like this setting of the emscripten define is duplicated in `emcc.py`? if we need it in both (but do we?), perhaps we could refactor it into a shared method?

juj replied:
I'm not sure, the machinery touches the old `EMCONFIGURE_JS` hacks that are used to support the `./configure` architecture. Looks like that mode does not have `__EMSCRIPTEN__` set, as it shouldn't(?).

kripken replied:
Yes, I see, the hacks there need to only add it in some cases.

I suppose we could have a `add_emscripten_define` method somewhere, which returns the one define or both depending on if we are in strict mode. How does that sound?

juj replied:
I'm not sure if adding that is much worth it, since this occurs exactly in two places, and it's quite lean to just add it in these locations. Eventually we'll be tearing down the code anyways when we get to remove `#define EMSCRIPTEN` altogether, so building a new abstraction for this seems temporary? Left this one as is for now, let me know if you feel strongly about this.

kripken commented on tools/shared.py:
```diff
@@ -471,6 +471,10 @@ def find_temp_directory():
 
+# Returns true if Emscripten is running in 'strict' mode, in which deprecated compiler features are not supported.
+def is_emscripten_strict():
+  return (os.environ.get('EMCC_STRICT') and int(os.environ.get('EMCC_STRICT')) != 0) or Settings.STRICT
```
this can try to access `Settings.STRICT` before the `Settings` object exists, that worries me a lot.

kripken replied:
In particular, we should have tests that check usage with both the env var and without, of things both working and not, so we hit that code path.

juj replied:
Hit that case, and had to adapt.

juj commented:
Pushed some new changes, namely made the `browser` suite to pass fully when running in `EMCC_STRICT` mode enabled. @kripken: could you take a fresh peek?

juj commented:
Updated the PR and rebased to latest. This should be good to go now.

kripken commented on emcc.py:
```diff
@@ -934,8 +939,16 @@ def detect_fixed_language_mode(args):
+        if not found:
+          system_js_libraries += shared.Building.path_to_system_js_libraries(lib)
+
+      # Certain linker flags imply some link libraries to be pulled in by default.
```
i still feel these lines are nicer in something like `shared.Building.add_default_system_libraries_from_settings_changes` (or a better name ;) ), but i don't feel strongly about it.

kripken replied:
Maybe leave that for a possible followup.

kripken commented on src/library_fs.js:
```diff
@@ -1,5 +1,15 @@
-  $FS__deps: ['$ERRNO_CODES', '$ERRNO_MESSAGES', '__setErrNo', '$PATH', '$TTY', '$MEMFS', '$IDBFS', '$NODEFS', '$WORKERFS', 'stdin', 'stdout', 'stderr'],
+  $FS__deps: ['$ERRNO_CODES', '$ERRNO_MESSAGES', '__setErrNo', '$PATH', '$TTY', '$MEMFS',
+#if __EMSCRIPTEN_HAS_idbfs_js__
+  '$IDBFS',
```
indentation not consistent with the following ones

juj replied:
Fixed

kripken commented on src/settings.js:
```diff
@@ -501,6 +505,22 @@ var WARN_ON_UNDEFINED_SYMBOLS = 1; // If set to 1, we will warn on any undefined
+                                    // building with -s ERROR_ON_MISSING_LIBRARIES=0 setting,
+                                    // prefer to set that option explicitly in your build system.
+
+var SYSTEM_JS_LIBRARIES = []; // Specifies a list of Emscripten-provided JS libraries to link against.
```
Currently we have `js_libraries` in `emcc.py`, which invokes `emscripten.py` with `--libraries`. It seems like `emcc.py` could use this new variable, and we could get rid of `--libraries`? That would avoid duplicated functionality.

kripken replied:
Although for people using `emscripten.py` directly, maybe we shouldn't remove that.

juj replied:
Hmm, that might be true, though not sure if I want to change this right now at least, there doesn't seem to be any tests for the `--libraries` flag in our test suite.

kripken commented on tools/shared.py:
```diff
@@ -472,6 +472,15 @@ def find_temp_directory():
 
+# Returns true if Emscripten is running in 'strict' mode, in which deprecated compiler features are not supported.
+def is_emscripten_strict():
+  if os.environ.get('EMCC_STRICT') and int(os.environ.get('EMCC_STRICT')) != 0: return True
```
this `int` operation may throw, if the value is e.g. `a`. instead, please do `os.environ.get('EMCC_STRICT') != '0'`

juj replied:
I think it's good to throw, no? That would indicate someone messed up setting the env. var, so a critical abort there would make sense, as opposed to e.g. silently treating `export EMCC_STRICT=a` as falsey or something similar.

kripken replied:
throwing might be ok, but it'll throw something like `ValueError: invalid literal for int() with base 10` which is not very clear. and in all other places, we accept `'0'` as "no", and everything else, including `"a"` as yes.

kripken commented on tools/shared.py:
```diff
@@ -927,12 +936,14 @@ def get_llvm_target():
                                    '-D__unix__']
 
+  # The preprocessor define EMSCRIPTEN is deprecated. Don't pass it to code in strict mode. Code should use the define __EMSCRIPTEN__ instead.
+  if not is_emscripten_strict(): COMPILER_OPTS += ['-DEMSCRIPTEN']
```
i think we discussed this already - this code happens before the `STRICT` setting is applied. so only the env var can influence this I think? ditto below.

kripken commented:
Some testing comments:

 * I see you got the test suite to pass with or without `STRICT`? Is there a test that verifies the old behavior, e.g. a GL test that doesn't `-lGL`?
 * We need direct tests for the strict mode features, checking both the env and the setting, in `test_other.py`.

juj commented:
> I see you got the test suite to pass with or without STRICT? Is there a test that verifies the old behavior, e.g. a GL test that doesn't -lGL?

Added test `browser.test_sdl1_in_emscripten_nonstrict_mode` to verify that code still passes with the old behavior.

> We need direct tests for the strict mode features, checking both the env and the setting, in test_other.py.

Added test `other.test_error_on_missing_libraries` to verify that the strict mode applies correctly on both command line and env. var, and that the defaults for strict mode for this specific flag are as expected.

juj commented:
When testing, I realize that the preprocessor define check goes all the way to https://github.com/kripken/emscripten-fastcomp-clang/pull/11

kripken commented:
Thanks, lgtm aside from that tiny test thing.

kripken commented:
Oh and the `'0'` thing.

juj commented:
Changed the '0' thing and rebased to latest. This would be good to merge in. Existing behavior should not be changed in any way, so should be safe to add in.

juj commented:
@kripken: Does this look good to you to merge? I'd really like to get this in to 1.37.0, because this adds the new `-s ERROR_ON_MISSING_LIBRARIES=0/1` option so people will have an ability to turn that one on already.

kripken commented on emcc.py:
```diff
@@ -918,6 +922,24 @@ def detect_fixed_language_mode(args):
       if separate_asm:
         shared.Settings.SEPARATE_ASM = os.path.basename(asm_target)
 
+      if 'EMCC_STRICT' in os.environ:
```
this happens after the code above for `-DEMSCRIPTEN`, so it seems it would only work for the setting and not the environment var?

kripken replied:
or wait, is the above for the configure stuff? then it's ok.

kripken commented on emcc.py:
```diff
@@ -918,6 +922,24 @@ def detect_fixed_language_mode(args):
+      if 'EMCC_STRICT' in os.environ:
+        shared.Settings.STRICT = os.environ.get('EMCC_STRICT') != '0'
+
+      STRICT = ([None] + filter(lambda x: x.startswith('STRICT='), settings_changes))[-1]
```
why `[None]`?

juj replied:
Hmm I guess it doesn't matter much, let me go with `False` instead since it's evaluated to true/false below.

kripken replied:
Sorry, I wasn't clear. My issue isn't with None vs False. It's that I don't understand that line - why does it need `[..something..]` at all? I guess it's so the `-1` can work, because we happen to just want the last one but the line just seems overcomplicated.

and we do this twice. how about adding a method, `get_last_change` which returns the last change, or None if there isn't one?

juj replied:
Ohh, I see. Refactored that to its own function, that looks better now, what do you think?

kripken commented on emcc.py:
```diff
@@ -918,6 +922,24 @@ def detect_fixed_language_mode(args):
+        shared.Settings.STRICT = os.environ.get('EMCC_STRICT') != '0'
+
+      STRICT = ([None] + filter(lambda x: x.startswith('STRICT='), settings_changes))[-1]
+      if STRICT:
```
I worry about creating a var called `STRICT`, might confuse us in the future as to what to use. How about `temp_strict` or `strict_changes`?

juj replied:
Ok, opting for `strict_cmdline` since the variable comes from a command line setting.

kripken commented on emcc.py:
```diff
@@ -918,6 +922,24 @@ def detect_fixed_language_mode(args):
+
+      STRICT = ([None] + filter(lambda x: x.startswith('STRICT='), settings_changes))[-1]
+      if STRICT:
+        shared.Settings.STRICT = int(STRICT[len('STRICT='):])
```
The settings changes would set `Settings.STRICT` for you - why do we even need to do this?

kripken replied:
Oh I see, because we need to do the library stuff first.

But then, let's comment that above here and not just below.

juj replied:
Check.

kripken commented on emcc.py:
```diff
@@ -918,6 +922,24 @@ def detect_fixed_language_mode(args):
+        shared.Settings.ERROR_ON_MISSING_LIBRARIES = 1
+
+      # Libraries are searched before settings_changes are applied, so pull its value from the command line already here.
+      ERROR_ON_MISSING_LIBRARIES = ([None] + filter(lambda x: x.startswith('ERROR_ON_MISSING_LIBRARIES='), settings_changes))[-1]
```
also, why `[None]`?

juj commented on emcc.py:
```diff
@@ -925,18 +925,23 @@ def detect_fixed_language_mode(args):
+      # Libraries are searched before settings_changes are applied, so apply the value for STRICT and ERROR_ON_MISSING_LIBRARIES from
+      # command line already now.
+
+      def last_element_or_none(arr):
```
Ahh ok, updated.
